Key Design of HDFS Architecture

RuthAMare opened the pull request:
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [ ] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+  - url: /engineering-education/key-design-of-hdfs-architecture/hero.jpg
+    alt: HDFS Architecture cover image 
+---
+HDFS (Hadoop Distributed File System) is a big data distributed file system storage by Apache that is implemented within the hadoop framework and that needs to have several features of design implemented to work effectively in processing, distibuting and storing big data.
```
```suggestion
HDFS (Hadoop Distributed File System) is a big data distributed file system storage by Apache. It is implemented within the Hadoop framework and that needs to have several features of design implemented to work effectively in processing, distributing, and storing big data.
```

srishilesh replied:
Please make use of [hemingwayapp](hemingwayapp.com) to write shorter and meaningful sentences.

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+
+### Overview
+This article will cover:
+- [ntroduction to Big Data Framework](#Introduction-to-Big-Data-framework)
```
```suggestion
- [Introduction to Big Data Framework](#Introduction-to-Big-Data-framework)
```

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+
+### Overview
+This article will cover:
+- [ntroduction to Big Data Framework](#Introduction-to-Big-Data-framework)
+- [General Design of HDFS architecture](#General-Design-of-HDFS-architecture)
+- [Configuring HDFS data storage policies](#Configuring-HDFS-data-storage-policies)
+- [Colocation and its benefits in HDFS](#Colocation-and-its-benefits-in-HDFS)
```
On the markdown headers part, you must write everything is lower case. Something like this:

`- [Introduction to Big Data Framework](#introduction-to-big-data-framework)`

Make the changes for the remaining headings.

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+- [Configuring HDFS data storage policies](#Configuring-HDFS-data-storage-policies)
+- [Colocation and its benefits in HDFS](#Colocation-and-its-benefits-in-HDFS)
+
+### 1) Introduction to Big Data Framework
```
Please remove the numbering.

If you put them here, them in the `Table of content` you have to make the changes accordingly like `#1-introduction-to-big-data-framework`

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+### 1) Introduction to Big Data Framework
+Big data is data in sets that are of high volume and complexity beyond what traditional data processing software applications can deal with.
+
+Big data framework is characterized by 4Vs namely; Variety (data is of various forms and types), Velocity (data processing speed is high), Value (low data value density), and Volume (massive amount of data).
```
Instead of writing the `4Vs` sequentially, write them as numbered lists.

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+
+This information is also synchronized between the Active and the Standby NameNode periodically.
+
+***Robustness:*** HDFS stores data reliably even when failures occur. Its robustness takes into account the 3 common types of failures: DataNode failure, NameNode failure, and network failures.
```
Convert the `3 types` to numbered lists

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+- Ensures metadata reliability - the transaction log mechanism is used to operate metadata, which is stored on both active and standby NameNodes. The snapshot mechanism of the file system ensures that data can be recovered promptly when a misoperation occurs.
+- Provides the security mode - HDFS provides a unique security mode to prevent a fault from spreading when a DataNode or hard disk faulty.
+
+***Data storage policy:*** HDFS supports 6 storage policies namely;
```
```suggestion
***Data storage policy:*** HDFS supports 6 storage policies namely:
```

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+- One_SSD – Storage of a single replica on SSD and other replicas on DISK.
+- All_SSD – Storage of all replicas on SSD.
+
+### 3) Configuring HDFS data storage policies
```
Remove `3)`

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+### 3) Configuring HDFS data storage policies
+The HDFS NameNode automatically selects DataNodes to store data replicas by default. This can be done by the following scenarios:
+
++ *Layered storage*
```
Instead of italicizing this, you can make this as another subheading with `####` as the header

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+
+Storage policies for different scenarios are formulated by combining the four types of storage devices.
+
++ *Tag storage*
+Select a proper DataNode according to directory tags indicate data importance levels.
+
++ *Node Group Storage*
+Store key data in highly reliable node groups because the DataNode cluster uses heterogeneous servers.
```
Do the same here too

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+This is implemented as a solution to the great consumption of network resources during a massive migration of data that affects the processing speed of massive data and system performance greatly.
+
+**Benefits of colocation**
++	Reduces network bandwidth and resource consumption.
++	Enhances easy and quick access to data.
```
For bulleted points, use `-` instead of `+`

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,120 @@
+
+***Data replication:*** In HDFS, a file’s blocks are replicated for fault tolerance and the number of replicas can be specified by an application. This can be done at creation time but is subject to change at will.
+
+***Metadata persistence:*** The HDFS NameNode stores the namespace. The NameNode consistently records every change that occurs in file system metadata in a transaction log file called the EditLog. Whenever a new file is created in HDFS, the NameNode inserts a record into the EditLog indicating the creation of the new file.
```
For `EditLog`, you could enclose them within ` `` `. Similarly, for any unique name or variable, you must enclose them within ` `` `.

Make this change wherever needed

srishilesh reviewed (changes requested):
Hey @RuthAMare,

Thanks for the PR.

This article looks amazing. I have requested a few changes. Please let me know when it's done, we can then move on to the tech review.

As a general suggestion, make use of [hemingwayapp](hemingwayapp.com) to write short and meaningful sentences.

Make sure go through each comment, and resolve them before requesting for the review.

Thank you.

RuthAMare commented:
> Hey @RuthAMare,
> 
> Thanks for the PR.
> 
> This article looks amazing. I have requested a few changes. Please let me know when it's done, we can then move on to the tech review.
> 
> As a general suggestion, make use of [hemingwayapp](hemingwayapp.com) to write short and meaningful sentences.
> 
> Make sure go through each comment, and resolve them before requesting for the review.
> 
> Thank you.

@srishilesh I've done the changes requested.
Thank you

srishilesh commented on articles/key-design-of-hdfs-architecture/index.md:
```diff
@@ -0,0 +1,133 @@
+
+***High Availability:*** HDFS supports features such as data replication which enhances the availability of the system. A single block of data is replicated in 3 nodes, so that even if a single node fail, a client can access the block from 2 other nodes.
+
+Data can still be accessed normally even when a failure occurs on the DataNode or NameNode.
```
Could you please explain what a `DataNode` and `NameNode` is?

srishilesh reviewed (changes requested):
Hey @RuthAMare,

Thank you for making the requested changes.

I have added 1 minor suggestion. Please resolve it before we move for the final review.

Thank you.

RuthAMare commented:
> Hey @RuthAMare,
> 
> Thank you for making the requested changes.
> 
> I have added 1 minor suggestion. Please resolve it before we move for the final review.
> 
> Thank you.

Thank you for the suggestion, i just added it.

srishilesh reviewed (approved):
Hey @RuthAMare,

Thank you for making the change.

Moving this to the final review.

Thank you.

hectorkambow commented:
programs check done

hectorkambow commented:
👋  @RuthAMare  
Overall a nice article - it seems to cover some good topics. 
Please go through my final edits - as I changed a couple things around - and I want to make sure we did not change the meaning of the sentences. 

I found the last sentence confusing as I felt new topics were being discussed. 

cc @srishilesh
